src/lib/ams.c

```diff
@@ -0,0 +1,748 @@
+		if ((producer_routing_table[iter].message_type_id == message_type_id) &&
+		    (producer_routing_table[iter].producer_instance_id == instance_id) &&
+		    (producer_routing_table[iter].producer_module_id == module_id)) {
+			producer_routing_table[iter].message_type_id = AMS_INVALID_MSG_TYPE;
+			break;
+		}
+	}
+	ams_release(shared_c);
+	return 0;
+}
+
+int ams_register_consumer(uint32_t message_type_id,
+			  uint16_t module_id,
+			  uint16_t instance_id,
+			  ams_msg_callback_fn function,
+			  void *ctx)
+{
+	struct async_message_service *ams = *arch_ams_get();
+	struct ams_routing_entry __sparse_cache *routing_table;
+	struct ams_shared_context __sparse_cache *shared_c;
+	int err = -EINVAL;
+
+	if (ams->ams_context == NULL || function == NULL)
+		return -EINVAL;
+
+	shared_c = ams_acquire(ams->ams_context->shared);
+
+	routing_table = shared_c->rt_table;
+	for (int iter = 0; iter < shared_c->routing_table_size; iter++) {
+		/* Search for first invalid entry */
+		if (routing_table[iter].message_type_id == AMS_INVALID_MSG_TYPE) {
+			/* Add entry to routing table for local service */
+			routing_table[iter].consumer_callback = function;
+			routing_table[iter].message_type_id = message_type_id;
+			routing_table[iter].consumer_instance_id = instance_id;
+			routing_table[iter].consumer_module_id = module_id;
+			routing_table[iter].consumer_core_id = cpu_get_id();
+			routing_table[iter].ctx = ctx;
+
+			/* Exit loop since we added new entry */
+			err = 0;
+			break;
+		}
+	}
+
+	ams_release(shared_c);
+	return err;
+}
+
+int ams_unregister_consumer(uint32_t message_type_id,
+			    uint16_t module_id,
+			    uint16_t instance_id,
+			    ams_msg_callback_fn function)
+{
+	struct async_message_service *ams = *arch_ams_get();
+	struct ams_routing_entry __sparse_cache *routing_table;
+	struct ams_shared_context __sparse_cache *shared_c;
+	int err = -EINVAL;
+
+	if (ams->ams_context == NULL)
+		return -EINVAL;
+
+	shared_c = ams_acquire(ams->ams_context->shared);
+	routing_table = shared_c->rt_table;
+	for (int iter = 0; iter < shared_c->routing_table_size; iter++) {
+		/* Search for required entry */
+		if ((routing_table[iter].message_type_id == message_type_id) &&
+		    (routing_table[iter].consumer_module_id == module_id) &&
+		    (routing_table[iter].consumer_instance_id == instance_id) &&
+		    (routing_table[iter].consumer_callback == function)) {
+			/* Remove this entry from routing table */
+			routing_table[iter].message_type_id = AMS_INVALID_MSG_TYPE;
+			routing_table[iter].consumer_callback = NULL;
+
+			/* Exit loop since we removed entry */
+			err = 0;
+			break;
+		}
+	}
+
+	ams_release(shared_c);
+	return err;
+}
+
+static uint32_t ams_push_slot(struct ams_shared_context __sparse_cache *ctx_shared,
+			      const struct ams_message_payload *msg,
+			      uint16_t module_id, uint16_t instance_id)
+{
+	int err;
+
+	for (uint32_t i = 0; i < ARRAY_SIZE(ctx_shared->slots); ++i) {
```

Comment:
why are you looking for **a** slot? Slots are per-core, why not just take **the** slot for the correct core? And why do we check `slot_uses` for 0? If slots are per core, then you have to use the respective core, and not pick up the next one if this one is used